server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -204,6 +220,29 @@ static boolean execute(SearchContext searchContext,
                 hasFilterCollector = true;
             }
 
+            // try to rewrite numeric or date sort to the optimized distanceFeatureQuery
+            if ((searchContext.sort() != null) && SYS_PROP_LONG_SORT_OPTIMIZED) {
+                try {
+                    Query rewrittenQuery = tryRewriteLongSort(searchContext, searcher.getIndexReader(), query, hasFilterCollector);
+                    if (rewrittenQuery != null) {
+                        query = rewrittenQuery;
+                        // modify sorts: add sort on _score as 1st sort, and move the sort on the original field as the 2nd sort
+                        SortField[] oldSortFields = searchContext.sort().sort.getSort();
+                        DocValueFormat[] oldFormats = searchContext.sort().formats;
+                        SortField[] newSortFields = new SortField[oldSortFields.length + 1];
+                        DocValueFormat[] newFormats = new DocValueFormat[oldSortFields.length + 1];
+                        newSortFields[0] = SortField.FIELD_SCORE;
+                        newFormats[0] = DocValueFormat.RAW;
+                        System.arraycopy(oldSortFields, 0, newSortFields, 1, oldSortFields.length);
+                        System.arraycopy(oldFormats, 0, newFormats, 1, oldFormats.length);
+                        sortAndFormatsForRewrittenNumericSort = searchContext.sort(); // stash SortAndFormats to restore it later
+                        searchContext.sort(new SortAndFormats(new Sort(newSortFields), newFormats));
+                    }
+                } catch (IOException e) {
+                    // in case of errors do nothing - keep the same query
```

Comment:
this doesn't look safe as we might have changed the query already when this exception is caught?

Reply:
@jpountz Can you please explain how this may happen? The way I see it is that the only way we can change the query if there is not exception in `tryRewriteLongSort`.

Reply:
Imagine the case that the exception is thrown after `query = rewrittenQuery;`. Maybe there is nothing that throws an exception today, but even then this would be fragile and this invariant could get broken by a seemingly safe refactoring.

Reply:
@jpountz Thanks, Adrien; makes sense. This addressed in the last commit, I have removed the `try` block.  I think it is reasonable if there are some IO exceptions during sort optimization, we can fail the whole search request.